scripts/gha/test_simulator.py

```diff
@@ -387,23 +389,77 @@ def _build_tvos_helper(helper_project, device_name, device_os):
         return os.path.join(file_dir, file_name)
 
 
-def _record_apple_tests(video_name):
-  command = "xcrun simctl io booted recordVideo -f --codec=h264 %s" % video_name
-  logging.info("Recording test: %s", command)
-  return subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, preexec_fn=os.setsid)
-
-
-def _stop_recording(record_process):
-  logging.info("Stop recording test")
-  try:
-    os.killpg(record_process.pid, signal.SIGINT)
-  except:
-    logging.info("Stop recording test failed!!!")
+def _record_apple_tests(video_name, display):
+  args = [
+    "xcrun",
+    "simctl",
+    "io",
+    "booted",
+    "recordVideo",
+    "-f",
+    "--codec=h264",
+  ]
+  if display is not None:
+    args.append("--display=" + display)
+  args.append(video_name)
+  return _start_recording(args)
+
+
+def _start_recording(args):
+  logging.info("Starting screen recording: %s", subprocess.list2cmdline(args))
+
+  output_file = tempfile.TemporaryFile()
+
+  # Specify CREATE_NEW_PROCESS_GROUP on Windows because it is required for
+  # sending CTRL_C_SIGNAL, which is done by _stop_recording().
+  proc = subprocess.Popen(
+    args,
+    stdout=output_file,
+    stderr=subprocess.STDOUT,
+    creationflags=
+        subprocess.CREATE_NEW_PROCESS_GROUP
+        if platform.system() == "Windows"
+        else 0,
+  )
+  logging.info("Started screen recording with PID %s", proc.pid)
+  return (proc, output_file)
+
+
+def _stop_recording(start_recording_retval):
+  (proc, output_file) = start_recording_retval
+  logging.info("Stopping screen recording with PID %s by sending CTRL+C",
+    proc.pid)
+
+  # Make sure that `returncode` is set on the `Popen` object.
+  proc.poll()
+
+  if proc.returncode is not None:
```

Comment:
nit: `if proc.returncode:`
Also, on Android Emulator, if the exit code is 0, there is no log: https://github.com/firebase/firebase-cpp-sdk/runs/8029138294?check_suite_focus=true#step:11:6091

Reply:
I intentionally checked for `is not None` here because I want to check if the process has completed. If returncode is None then the process is still alive, which is the expected behavior. If returncode is any number, even zero, then that means that the process terminated prematurely and I want to echo whatever output there is to assist with any debugging.

In the link you provided, I can't tell which commit that integration test run was executed at, but it looks like it was from a run where there was a bug where I forgot to call proc.poll() to set returncode (which was fixed by https://github.com/firebase/firebase-cpp-sdk/pull/1058/commits/962e71e6c65e13ba48d62e6328a5679dd99c8fcf). Now, with the fix, it won't print the screenrecord output if the process is still in progress.